File: easyplot/scripts/easyplot_dongfang.py
"""Opt-in, source-attributed Dongfang colour schemes. No network at runtime."""
from __future__ import annotations

import json
from numbers import Integral
from pathlib import Path

import numpy as np
from matplotlib.colors import ListedColormap

REGISTRY_PATH = Path(__file__).resolve().parents[1] / "assets" / "dongfang.json"


def registry():
    """Return a fresh registry including source identities and curation rationale."""
    return json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))


def dongfang_color(name: str) -> str:
    for colour in registry()["source_colours"]:
        if name == colour["name"] or name in colour["aliases"]:
            return colour["hex"]
    raise ValueError(f"Unknown Dongfang source colour: {name!r}")


def dongfang_palettes():
    return {name: {key: value for key, value in entry.items() if key in
                  ("label", "type", "max_n", "rationale", "uses")}
            for name, entry in registry()["palettes"].items()}


def dongfang_palette(name="danqing", n=None, role="fill", reverse=False):
    """Stable category prefixes or samples of a frozen continuous LUT.

    Map the returned values to explicit group names once per figure. For
    diverging data, declare a meaningful centre in the plotting normalization.
    """
    palettes = registry()["palettes"]
    if not isinstance(reverse, (bool, np.bool_)):
        raise ValueError("reverse must be a boolean")
    if name not in palettes:
        raise ValueError(f"Unknown palette {name!r}; choose from {', '.join(palettes)}")
    spec = palettes[name]
    if role not in ("fill", "line") or role not in spec:
        raise ValueError(f"Role {role!r} is unavailable for {name}")
    colours = spec[role]
    if n is None:
        n = len(colours)
    if isinstance(n, bool) or not isinstance(n, Integral) or n < 1:
        raise ValueError("n must be a positive integer")
    if spec["type"] == "qualitative":
        if n > spec["max_n"]:
            raise ValueError(f"{name} supports at most {spec['max_n']} categories; use facets or another scheme")
        result = colours[:n]
    else:
        indices = [len(colours) // 2] if n == 1 else np.floor(np.linspace(0, len(colours) - 1, n) + 0.5).astype(int)
        result = [colours[index] for index in indices]
    return result[::-1] if reverse else result


def dongfang_cmap(name="qingci", reverse=False):
    spec = registry()["palettes"].get(name)
    if spec is None or spec["type"] == "qualitative":
        raise ValueError("A continuous cmap requires qingci, qinglan, qingzhu or moyun")
    cmap = ListedColormap(dongfang_palette(name, reverse=reverse), name=f"dongfang_{name}")
    cmap.set_bad(registry()["neutrals"]["missing"])
    return cmap

File: easyplot/scripts/easyplot_dongfang.R
# Opt-in Dongfang palettes. Source this file; the existing EasyPlot runtime
# and defaults are unchanged. Name returned colours by semantic group once.

local({
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    stop("Dongfang palettes require jsonlite.", call. = FALSE)
  }
  source_frames <- Filter(function(frame) !is.null(frame$ofile), sys.frames())
  if (!length(source_frames)) {
    stop("Load easyplot_dongfang.R with source().", call. = FALSE)
  }
  source_frame <- source_frames[[length(source_frames)]]
  script_path <- if (isTRUE(source_frame$chdir)) basename(source_frame$ofile) else source_frame$ofile
  script_path <- normalizePath(script_path, mustWork = TRUE)
  registry <- jsonlite::fromJSON(
    file.path(dirname(script_path), "..", "assets", "dongfang.json"),
    simplifyVector = FALSE
  )
  colours <- unlist(lapply(registry$source_colours, function(colour) {
    keys <- c(colour$name, unlist(colour$aliases, use.names = FALSE))
    stats::setNames(rep(colour$hex, length(keys)), keys)
  }))
  check_name <- function(name, known, kind) {
    if (!is.character(name) || length(name) != 1L || is.na(name) ||
        !nzchar(trimws(name)) || !name %in% known) {
      stop("Unknown or empty Dongfang ", kind, " name.", call. = FALSE)
    }
  }

  dongfang_color <- function(name) {
    check_name(name, names(colours), "source colour")
    unname(colours[[name]])
  }

  dongfang_palette <- function(name = "danqing", n = NULL, role = "fill", reverse = FALSE) {
    check_name(name, names(registry$palettes), "palette")
    spec <- registry$palettes[[name]]
    if (!is.character(role) || length(role) != 1L || is.na(role) ||
        !role %in% c("fill", "line") || !role %in% names(spec)) {
      stop("Role must be 'fill', or 'line' for a qualitative palette.", call. = FALSE)
    }
    if (!is.logical(reverse) || length(reverse) != 1L || is.na(reverse)) {
      stop("reverse must be TRUE or FALSE.", call. = FALSE)
    }
    values <- unlist(spec[[role]], use.names = FALSE)
    if (is.null(n)) n <- length(values)
    if (!typeof(n) %in% c("integer", "double") || length(n) != 1L ||
        !is.finite(n) || n < 1 || n != floor(n)) {
      stop("n must be one positive integer.", call. = FALSE)
    }
    if (spec$type == "qualitative") {
      if (n > spec$max_n) {
        stop(name, " supports at most ", spec$max_n, " categories.", call. = FALSE)
      }
      values <- values[seq_len(n)]
    } else {
      indices <- if (n == 1) 129L else floor(seq(0, 256, length.out = n) + 0.5) + 1L
      values <- values[indices]
    }
    if (reverse) rev(values) else values
  }

  dongfang_palettes <- function() {
    lapply(registry$palettes, function(spec) {
      spec[c("label", "type", "max_n", "rationale", "uses")]
    })
  }

  # Only the three public helpers enter the sourcing environment. R's value
  # semantics isolate returned vectors/lists; locked bindings protect the cache.
  for (name in c("dongfang_color", "dongfang_palette", "dongfang_palettes")) {
    assign(name, get(name), envir = parent.env(environment()))
  }
  lockEnvironment(environment(), bindings = TRUE)
})
